libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt

```diff
@@ -62,11 +63,52 @@ abstract class ParentNode<NavTarget : Any>(
     plugins = plugins + navModel + childAware
 ), Resolver<NavTarget> {
 
-    private val permanentNavModel = PermanentNavModel<NavTarget>(
-        savedStateMap = buildContext.savedStateMap,
-        key = KEY_PERMANENT_NAV_MODEL,
-    )
-    val navModel: NavModel<NavTarget, *> = permanentNavModel + navModel
+    private lateinit var permanentNavModel: PermanentNavModel<NavTarget>
+    private var isPermanentNavModelProvided: Boolean? = null
+    val navModel: NavModel<NavTarget, *> = resolveNavModels(navModel, buildContext)
+
+    /**
+     * If PermanentNavModel is provided in the constructor, it will be retrieved and used. Original NavModel
+     * will not be modified. Otherwise, PermanentNavModel will be created internally, and combined with
+     * the original NavModel. In ths case it's not added to the list of plugins, and we'll have to call
+     * onSavedInstanceState manually.
+     */
+    private fun resolveNavModels(
+        navModel: NavModel<NavTarget, *>,
+        buildContext: BuildContext,
+    ): NavModel<NavTarget, *> {
+
+        val existingPermanentNavModel = retrievePermanentNavModel(navModel)
+        isPermanentNavModelProvided = existingPermanentNavModel != null
+
+        return if (existingPermanentNavModel != null) {
+            permanentNavModel = existingPermanentNavModel
+            navModel
+        } else {
+            permanentNavModel = PermanentNavModel(
+                savedStateMap = buildContext.savedStateMap,
+                key = KEY_PERMANENT_NAV_MODEL,
+            )
+            navModel + permanentNavModel
+        }
+    }
+
+    private fun retrievePermanentNavModel(navModel: NavModel<NavTarget, *>): PermanentNavModel<NavTarget>? =
+        when (navModel) {
+            is CombinedNavModel<NavTarget> -> {
+                navModel
+                    .navModels
+                    .find { it is PermanentNavModel<NavTarget> } as PermanentNavModel<NavTarget>?
```

Comment:
It seems like it's possible (though maybe not advised?) that a developer could nest a `CombinedNavModel` within a `CombinedNavModel`.

Perhaps this case should be covered as well? (if so, then the init block for the CombinedNavModel might need to be updated).

If we don't want to allow nesting `CombinedNavModel` within `CombinedNavModel` perhaps the `init` block of `CombinedNavModel` could prevent this as well? (though this hypothetically could be a breaking change.

i.e.

```
CombinedNavModel(
    listOf(
         CombinedNavModel(
             listOf(
                 PermanentNavModel()
             )
         )
    )
)
```

Reply:
yeah, I think we can prevent nesting of `CombinedNavModel` as it doesn't make much sense and harder to support

Reply:
It makes sense at the current moment.
First you pass `CombinedNavModel` via `ParentNode(navModel = backStack1 + backStack2)` then this `CombinedNavModel` is merged with internal `PermanentNavModel` into another `CombinedNavModel`. So at the end we have `CombinedNavModel` inside `CombinedNavModel` is a pretty basic case.

Reply:
@KovalevAndrey it sounds like this might be wise to keep? otherwise you should definitely update the changelog to call this out as a breaking change.

Reply:
@LachlanMcKee Looking at the code I'm leaning towards removing the `PermanentNavModel` from parentNode at all and use it the same way as any other navModels

Reply:
> Nesting makes sense at the current moment. First you pass `CombinedNavModel` via `ParentNode(navModel = backStack1 + backStack2)` then this `CombinedNavModel` is merged with internal `PermanentNavModel` into another `CombinedNavModel`. So at the end we have `CombinedNavModel` inside `CombinedNavModel` is a pretty basic case.

If we're using the provided API it doesn't get nested

<img width="813" alt="Screenshot 2023-09-30 at 10 48 19" src="https://github.com/bumble-tech/appyx/assets/5773436/b49019a4-011b-48fb-b992-88b86aa9c447">